[memory profiling] C++ tracing support

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96281
* #95753
* #95541
* __->__ #95357
* #96280
* #95356

Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.

zdevito committed:
[memory profiling][WIP] C++ tracing support

Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.

TODO:
* testing code apart from generating traces with memory profiler
* fallback to `backward` on non-x86 architectures
* symbolize needs to ensure pipes to addr2line do not get full

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling][WIP] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.

TODO:
* testing code apart from generating traces with memory profiler
* fallback to `backward` on non-x86 architectures
* symbolize needs to ensure pipes to addr2line do not get full

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling][WIP] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.

TODO:
* testing code apart from generating traces with memory profiler
* fallback to `backward` on non-x86 architectures
* symbolize needs to ensure pipes to addr2line do not get full

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling][WIP] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.

TODO:
* testing code apart from generating traces with memory profiler
* fallback to `backward` on non-x86 architectures
* symbolize needs to ensure pipes to addr2line do not get full

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling][WIP] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.

TODO:
* testing code apart from generating traces with memory profiler
* fallback to `backward` on non-x86 architectures
* symbolize needs to ensure pipes to addr2line do not get full

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling][WIP] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.

TODO:
* testing code apart from generating traces with memory profiler
* fallback to `backward` on non-x86 architectures
* symbolize needs to ensure pipes to addr2line do not get full

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling][WIP] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.

TODO:
* testing code apart from generating traces with memory profiler
* fallback to `backward` on non-x86 architectures
* symbolize needs to ensure pipes to addr2line do not get full

[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.


[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.


[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.


[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.


[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.


[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.


[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.


[ghstack-poisoned]

bertmaher reviewed (approved):
I spent this morning reviewing, and everything I see looks good!  It'd be nice to have some links to relevant specs, as you said.  Some things are weirdly hard to find (others pop up right away, but yeah).  The eh_frame documentation, Dwarf spec, and *maybe* elf spec (although you can kind of get away with just knowing that eh_frame sections exist)

zdevito commented on torch/csrc/profiler/unwind/unwind.cpp:
```diff
@@ -0,0 +1,371 @@
+        return e;
+      }
+    }
+    const char* args[] = {
+        "addr2line", "-C", "-f", "-e", info.name().c_str(), nullptr};
+    entries.emplace_back(
+        Entry{&info, std::make_unique<Communicate>("addr2line", args), {}});
```
I have a companion patch that uses the internal symbolizer code when this runs in fbcode but for OSS this seems like the most robust way to do it since it is so widely installed.

zdevito committed:
Update on "[memory profiling] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.


[ghstack-poisoned]

zdevito committed:
Update on "[memory profiling] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.


[ghstack-poisoned]

zdevito commented:
@pytorchbot merge

zdevito committed:
Update on "[memory profiling] C++ tracing support"


Adds the ability to quickly generate stack traces for C++,
and combine Python, TorchScript, and C++ frames into a single trace.

This makes it possible for the memory tracer to record allocations inside
C++ code (e.g. convolution temporaries, backward operators).

The unwinder code is ~10x faster than execinfo.h's backward because it
cache fast unwinder routines for instruction pointers that have already been seen.
It is also only 1.2--2x slower than copying the entire stack (the approach perf takes),
while using 2 orders of magnitude less space per stack.


[ghstack-poisoned]

zdevito commented:
@pytorchbot merge
